pkg/agent/events/handlers/cputhrottle/utils.go

```diff
@@ -0,0 +1,142 @@
+		return h.getDefaultCPUQuota(pod), nil
+	}
+
+	quota, err := strconv.ParseInt(quotaStr, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse CPU quota: %v", err)
+	}
+
+	return quota, nil
+}
+
+// getDefaultCPUQuota calculates the CPU quota based on the original Pod specification,
+// This represents the intended CPU limit as defined in the Pod's resource requirements,
+// before any runtime modifications like throttling.
+func (h *CPUThrottleHandler) getDefaultCPUQuota(pod *v1.Pod) int64 {
+	// Calculate default cpu quota based on containers' cpu limits.
+	// CPU limits in Kubernetes are specified in CPU units(cores), where:
+	// - 1 core = 1000 millicores = 1000m
+	// - The quota is calculated as: (millicores * period) / 1000
+	// For example: 2 cores = 2000m = (2000 * 100000) / 1000 = 200000 microseconds
+	for _, container := range pod.Spec.Containers {
+		if cpuLimit := container.Resources.Limits.Cpu(); cpuLimit != nil {
+			cpuMillis := cpuLimit.MilliValue()
+			quota := cpuMillis * CPUPeriod / 1000
+			return quota
+		}
+	}
+
+	// If cpu has no limit, return default as 1 CPU core
+	return 1 * CPUPeriod
+}
+
+// calculateSteppedQuota computes the new CPU quota after applying throttling step.
+// The throttling uses a gradual approach to avoid sudden performance drops that
+// could destabilize running applications.
+func (h *CPUThrottleHandler) calculateSteppedQuota(pod *v1.Pod, currentQuota int64) int64 {
+	stepPercent := h.throttleStepPercent
+	// Calculate the quota for the tiered limit
+	stepReduction := currentQuota * int64(stepPercent) / 100
+	newQuota := currentQuota - stepReduction
+
+	// Calculate the minimum quota to prevent complete CPU starvation.
+	minQuota := h.calculateMinCPUQuota(pod)
+
+	if newQuota < minQuota {
+		return minQuota
+	}
+
+	return newQuota
+}
+
+// calculateMinCPUQuota determines the minimum CPU quota that should be preserved
+// This handler aims to throttle pods to a configurable percentage of their original limits
+func (h *CPUThrottleHandler) calculateMinCPUQuota(pod *v1.Pod) int64 {
+	// Always use percentage of the original CPU limit as the minimum quota
+	// This ensures consistent throttling behavior regardless of requests
+	// TODO: should respect resources.requests?
+	originalQuota := h.getDefaultCPUQuota(pod)
+
+	minPercent := int64(h.minCPUQuotaPercent)
+
+	minQuota := originalQuota * minPercent / 100
+
+	// Ensure we have an absolute minimum (at least 20% of one CPU core)
+	absoluteMin := int64(CPUPeriod * 20 / 100)
+	if minQuota < absoluteMin {
+		minQuota = absoluteMin
+	}
+
+	return minQuota
+}
+
+// applyCPUQuota writes the new CPU quota to the cgroup filesystem to enforce
+// the CPU limit at the kernel level. This is how Kubernetes actually implements
+// CPU resources.
+func (h *CPUThrottleHandler) applyCPUQuota(pod *v1.Pod, quota int64) (string, error) {
+	cgroupPath, err := h.cgroupMgr.GetPodCgroupPath(pod.Status.QOSClass, cgroup.CgroupCpuSubsystem, pod.UID)
```

Comment:
Actually, we just need to suppress the BE pods, therefore we don't need to get the cgroup path according to the pod's QoSClass, we just need to set the cpu quota for the BE pods' root cgroup '/sys/fs/cgroup/cpu/kubepods.slice/kubepods-besteffort.slice' (take systemd as an example)